File: apps/api/app/schemas/identity.py
"""Identity and Style schemas."""
from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class IdentityGraphResponse(BaseModel):
    """Schema for identity graph response - includes all fields from model."""

    id: UUID
    profile_id: UUID
    
    # Core identity
    current_role: Optional[str] = None
    industry: Optional[str] = None
    
    # Professional details
    expertise_areas: list[str] = Field(default_factory=list)
    career_highlights: list[str] = Field(default_factory=list)
    bio_summary: Optional[str] = None
    
    # Brand Strategy
    target_audience: Optional[str] = None
    desired_positioning: Optional[str] = None
    unique_angles: list[str] = Field(default_factory=list)
    aspirations: Optional[str] = None
    goals: Optional[str] = None
    
    # Personality & Content
    interests: list[str] = Field(default_factory=list)
    beliefs: list[str] = Field(default_factory=list)

    # Deep Identity (auto-extracted from LinkedIn posts)
    stories: list[Any] = Field(default_factory=list)
    opinion_statements: list[str] = Field(default_factory=list)
    interest_details: dict[str, str] = Field(default_factory=dict)

    # Content Strategy
    content_pillars: list[str] = Field(default_factory=list)
    narrative_themes: list[str] = Field(default_factory=list)

    # Credibility markers
    authority_angles: list[str] = Field(default_factory=list)

    # Metadata
    completeness_score: int = 0
    version: int
    last_updated_at: datetime
    created_at: datetime

    class Config:
        from_attributes = True


class IdentityGraphUpdate(BaseModel):
    """Schema for updating identity graph - all editable fields."""

    # Core identity
    current_role: Optional[str] = None
    industry: Optional[str] = None
    
    # Professional details
    expertise_areas: Optional[list[str]] = None
    career_highlights: Optional[list[str]] = None
    bio_summary: Optional[str] = None
    
    # Brand Strategy
    target_audience: Optional[str] = None
    desired_positioning: Optional[str] = None
    unique_angles: Optional[list[str]] = None
    aspirations: Optional[str] = None
    goals: Optional[str] = None
    
    # Personality & Content
    interests: Optional[list[str]] = None
    beliefs: Optional[list[str]] = None

    # Deep Identity (auto-extracted from LinkedIn posts)
    stories: Optional[list[Any]] = None
    opinion_statements: Optional[list[str]] = None
    interest_details: Optional[dict[str, str]] = None

    # Content Strategy
    content_pillars: Optional[list[str]] = None
    narrative_themes: Optional[list[str]] = None

    # Credibility markers
    authority_angles: Optional[list[str]] = None



class ToneSliders(BaseModel):
    """Schema for tone sliders."""

    formal_casual: float = Field(0.5, ge=0, le=1)
    technical_simple: float = Field(0.5, ge=0, le=1)
    serious_playful: float = Field(0.5, ge=0, le=1)
    humble_confident: float = Field(0.5, ge=0, le=1)


class FormatPreferences(BaseModel):
    """Schema for format preferences."""

    post: float = Field(0.5, ge=0, le=1)
    thread: float = Field(0.3, ge=0, le=1)
    carousel: float = Field(0.2, ge=0, le=1)


class StyleProfileResponse(BaseModel):
    """Schema for style profile response."""

    id: UUID
    profile_id: UUID
    tone_sliders: dict[str, float]
    format_preferences: dict[str, float]
    taboo_list: list[str] = Field(default_factory=list)
    preferred_hooks: list[str] = Field(default_factory=list)
    weights: dict[str, Any] = Field(default_factory=dict)
    version: int
    created_at: datetime
    updated_at: datetime
    writing_samples_count: int = 0

    class Config:
        from_attributes = True


class StyleProfileUpdate(BaseModel):
    """Schema for updating style profile."""

    tone_sliders: Optional[dict[str, float]] = None
    format_preferences: Optional[dict[str, float]] = None
    taboo_list: Optional[list[str]] = None
    preferred_hooks: Optional[list[str]] = None


class TimelineEventResponse(BaseModel):
    """Schema for timeline event."""
    id: UUID
    timeline_id: UUID
    title: str
    description: Optional[str] = None
    event_type: str
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    is_current: bool
    emotional_core: Optional[str] = None
    lessons_learned: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    source: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class TimelineEventUpdate(BaseModel):
    """Schema for updating a timeline event (enrichment from identity page)."""
    title: Optional[str] = None
    description: Optional[str] = None
    emotional_core: Optional[str] = None
    lessons_learned: Optional[list[str]] = None
    tags: Optional[list[str]] = None


class TimelineResponse(BaseModel):
    """Schema for timeline."""
    id: UUID
    identity_graph_id: UUID
    primary_focus: Optional[str] = None
    narrative_arc: Optional[str] = None
    events: list[TimelineEventResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: apps/web/src/lib/api/identity.ts
import { apiClient } from './client';

// ============================================
// Core Identity Types
// ============================================

export type TimelineEventType = 'education' | 'work' | 'usage' | 'achievement' | 'failure' | 'life_event' | 'pivot' | 'other';

export interface TimelineEvent {
  id: string;
  timeline_id: string;
  title: string;
  description: string | null;
  event_type: TimelineEventType;
  start_date: string | null;
  end_date: string | null;
  is_current: boolean;
  emotional_core: string | null;
  lessons_learned: string[];
  tags: string[];
  source: string | null;
  created_at: string;
}

export interface Timeline {
  id: string;
  identity_graph_id: string;
  primary_focus: string | null;
  narrative_arc: string | null;
  events: TimelineEvent[];
  created_at: string;
  updated_at: string;
}

export interface IdentityGraph {
  id: string;
  profile_id: string;

  // Core identity
  current_role: string | null;
  industry: string | null;

  // Professional details
  expertise_areas: string[];
  career_highlights: string[];
  bio_summary: string | null;

  // Brand Strategy
  target_audience: string | null;
  desired_positioning: string | null;
  unique_angles: string[];
  aspirations: string | null;
  goals: string | null;

  // Personality & Content
  interests: string[];
  beliefs: string[];

  // Deep Identity
  stories: any[];
  opinion_statements: string[];
  interest_details: Record<string, string>;

  // Content Strategy
  content_pillars: string[];
  narrative_themes: string[];

  // Credibility markers
  authority_angles: string[];

  // Metadata
  completeness_score: number;
  version: number;
  last_updated_at: string;
  created_at: string;

  // Relations
  timeline?: Timeline;
}

export interface IdentityGraphUpdate {
  // Core identity
  current_role?: string;
  industry?: string;

  // Professional details
  expertise_areas?: string[];
  career_highlights?: string[];
  bio_summary?: string;

  // Brand Strategy
  target_audience?: string;
  desired_positioning?: string;
  unique_angles?: string[];
  aspirations?: string;
  goals?: string;

  // Personality & Content
  interests?: string[];
  beliefs?: string[];

  // Deep Identity
  stories?: any[];
  opinion_statements?: string[];
  interest_details?: Record<string, string>;

  // Content Strategy
  content_pillars?: string[];
  narrative_themes?: string[];

  // Credibility markers
  authority_angles?: string[];
}

// ============================================
// Style Profile Types
// ============================================

export interface ToneSliders {
  formal_casual: number;
  technical_simple: number;
  serious_playful: number;
  humble_confident: number;
}

export interface FormatPreferences {
  post: number;
  thread: number;
  carousel: number;
}

export interface StyleProfile {
  id: string;
  profile_id: string;
  tone_sliders: ToneSliders;
  format_preferences: FormatPreferences;
  taboo_list: string[];
  preferred_hooks: string[];
  weights: Record<string, any>;
  version: number;
  created_at: string;
  updated_at: string;
  writing_samples_count?: number;
}

export interface StyleProfileUpdate {
  tone_sliders?: Partial<ToneSliders>;
  format_preferences?: Partial<FormatPreferences>;
  taboo_list?: string[];
  preferred_hooks?: string[];
}

// ============================================
// Identity Universe Types (Unified)
// ============================================

export interface PersonaData {
  persona_prompt: string | null;
  persona_prompt_updated_at: string | null;
  learned_preferences: string | null;
  learned_preferences_updated_at: string | null;
}

export interface IdentityUniverse {
  identity_graph: IdentityGraph;
  style_profile: StyleProfile | null;
  persona: PersonaData;
  profile_id: string;
  profile_name: string;
  location: string[] | null;
  completeness_score: number;
}

export interface RegenerationField {
  field_name: string;
  current_value: any;
  proposed_value: any;
  field_category: 'identity' | 'style' | 'persona';
}

export interface RegenerationPreview {
  profile_id: string;
  changes: RegenerationField[];
  new_persona_prompt: string | null;
  regeneration_scope: 'full' | 'partial';
}

export interface RegenerationRequest {
  scope: 'full' | 'partial';
  fields_to_regenerate?: string[];
}

// ============================================
// API Client
// ============================================

export const identityApi = {
  // Legacy endpoints
  getIdentityGraph: async (profileId: string): Promise<IdentityGraph> => {
    const response = await apiClient.get(`/profiles/${profileId}/identity-graph`);
    return response.data;
  },

  updateIdentityGraph: async (
    profileId: string,
    data: IdentityGraphUpdate
  ): Promise<IdentityGraph> => {
    const response = await apiClient.put(`/profiles/${profileId}/identity-graph`, data);
    return response.data;
  },

  getStyleProfile: async (profileId: string): Promise<StyleProfile> => {
    const response = await apiClient.get(`/profiles/${profileId}/style-profile`);
    return response.data;
  },

  updateStyleProfile: async (
    profileId: string,
    data: StyleProfileUpdate
  ): Promise<StyleProfile> => {
    const response = await apiClient.put(`/profiles/${profileId}/style-profile`, data);
    return response.data;
  },

  // Unified Identity Universe endpoints
  getIdentityUniverse: async (profileId: string): Promise<IdentityUniverse> => {
    const response = await apiClient.get(`/profiles/${profileId}/identity-universe`);
    return response.data;
  },

  previewRegeneration: async (
    profileId: string,
    request: RegenerationRequest
  ): Promise<RegenerationPreview> => {
    const response = await apiClient.post(
      `/profiles/${profileId}/identity-universe/regenerate-preview`,
      request
    );
    return response.data;
  },

  getTimeline: async (profileId: string): Promise<Timeline> => {
    const response = await apiClient.get(`/profiles/${profileId}/timeline`);
    return response.data;
  },

  updateTimelineEvent: async (
    profileId: string,
    eventId: string,
    data: { title?: string; description?: string; emotional_core?: string; lessons_learned?: string[]; tags?: string[] }
  ): Promise<TimelineEvent> => {
    const response = await apiClient.put(`/profiles/${profileId}/timeline/events/${eventId}`, data);
    return response.data;
  },
};
